File: (8 kyu) Square(n) Sum/(8 kyu) Square(n) Sum.py
# 1 Plain solution
def square_sum(numbers):
    sum = 0
    
    for x in numbers:
        sum += x * x
        
    return sum

# 2 Optimized solution
def square_sum(numbers):
    res = []
    for x in numbers:
        res.append(x**2)
    return(sum(res))

# 3 Clever solution
def square_sum(numbers):
    return sum(x ** 2 for x in numbers)

# 4 Coding golf
square_sum = lambda n: sum(x**2 for x in n)

File: (8 kyu) Square(n) Sum/(8 kyu) Square(n) Sum.rs
fn square_sum(vec: Vec<i32>) -> i32 {
    // #1
    // let mut res = 0;
    // for i in 0..vec.len() {
    //     res += vec[i] * vec[i]
    // }
    // res
    // #2
    // vec.iter().fold(0, |sum, x| sum + (x * x))
    // #3
    vec.iter().map(|x| x * x).sum()
}
